### ISSUE TITLE
Encoding byte strings in Python 3 causes infinite recursion

### ISSUE THREAD

**Issue by aldanor** [jsonpickle/jsonpickle#108]
```python
>>> jsonpickle.encode(b'foo')
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-5e089e18bf41> in <module>()
----> 1 jsonpickle.encode(b'foo')

/Users/ivansmirnov/dev/jsonpickle/jsonpickle/__init__.py in encode(value, unpicklable, make_refs, keys, max_depth, backend, warn, max_iter)
    129                           keys=keys,
    130                           max_depth=max_depth,
--> 131                           warn=warn)
    132 
    133 

/Users/ivansmirnov/dev/jsonpickle/jsonpickle/pickler.py in encode(value, unpicklable, make_refs, keys, max_depth, reset, backend, warn, context, max_iter)
     39                           warn=warn,
     40                           max_iter=max_iter)
---> 41     return backend.encode(context.flatten(value, reset=reset))
     42 
     43 

/Users/ivansmirnov/dev/jsonpickle/jsonpickle/pickler.py in flatten(self, obj, reset)
    139         if reset:
    140             self.reset()
--> 141         return self._flatten(obj)
    142 
    143     def _flatten(self, obj):

/Users/ivansmirnov/dev/jsonpickle/jsonpickle/pickler.py in _flatten(self, obj)
    143     def _flatten(self, obj):
    144         self._push()
--> 145         return self._pop(self._flatten_obj(obj))
    146 
    147     def _flatten_obj(self, obj):

/Users/ivansmirnov/dev/jsonpickle/jsonpickle/pickler.py in _flatten_obj(self, obj)
    159             return None
    160 
--> 161         return flatten_func(obj)
    162 
    163     def _list_recurse(self, obj):

/Users/ivansmirnov/dev/jsonpickle/jsonpickle/pickler.py in _ref_obj_instance(self, obj)
    214             # We've never seen this object so return its
    215             # json representation.
--> 216             return self._flatten_obj_instance(obj)
    217         # We've seen this object before so place an object
    218         # reference tag in the data. This avoids infinite recursion

...

/Users/ivansmirnov/dev/jsonpickle/jsonpickle/pickler.py in _flatten(self, obj)
    143     def _flatten(self, obj):
    144         self._push()
--> 145         return self._pop(self._flatten_obj(obj))
    146 
    147     def _flatten_obj(self, obj):

/Users/ivansmirnov/dev/jsonpickle/jsonpickle/pickler.py in _flatten_obj(self, obj)
    147     def _flatten_obj(self, obj):
    148         self._seen.append(obj)
--> 149         max_reached = self._depth == self._max_depth
    150 
    151         if max_reached or (not self.make_refs and id(obj) in self._objs):

RuntimeError: maximum recursion depth exceeded in comparison
```

**Comment by marcintustin** [comment]
Ouch. You're doing yeoman's work on finding our bugs!

-- 
Marcin Tustin
Tel: +44 (0) 7773 787 105 (UK)
       +1  917 553 3974 (US)

**Comment by aldanor** [comment]
This is a pretty bad one since this is a very basic built in type :) Wonder if it's a regression of a sort

**Comment by aldanor** [comment]
My guess is this:
```python
>>> b'foo'.__getnewargs__()
(b'foo',)
```

**Comment by aldanor** [comment]
The question is -- how do we want to handle bytestrings in general? Introduce a `tags.BYTES` and encode/decode the payload to/from UTF-8?

http://www.diveintopython3.net/strings.html#byte-arrays

**Comment by aldanor** [comment]
@davvid

**Comment by marcintustin** [comment]
That sounds right to me - it's a primitive type, after all.

**Comment by aldanor** [comment]
Idk, could be more compact or could be longer (I guess the main benefit using something like latin1 is that any ascii text would still remain readable).
```python
>>> import json
>>> s = b'\x00\x00\x00\x00\x00\x00\xf0?foo'
>>> json.dumps(s.decode('latin1'))
'"\\u0000\\u0000\\u0000\\u0000\\u0000\\u0000\\u00f0?foo"'
>>> import array
>>> json.dumps(list(array.array('B', s)))
'[0, 0, 0, 0, 0, 0, 240, 63, 102, 111, 111]'
```

**Comment by marcintustin** [comment]
Base64?

-- 
Marcin Tustin
Tel: +44 (0) 7773 787 105 (UK)
       +1  917 553 3974 (US)

**Comment by aldanor** [comment]
```python
>>> import base64
>>> base64.b64encode(s).decode('ascii')
'AAAAAAAA8D9mb28='
```

**Comment by marcintustin** [comment]
I was thinking more u'ascii'.encode('base64'), which seems to work fine,
and result in minimal expansion. Decimal coding with spaces and commas is
going to be suboptimal.

On Friday, January 9, 2015, Ivan Smirnov <notifications@github.com> wrote:

> >>> import base64>>> base64.b64encode(s).decode('ascii')'AAAAAAAA8D9mb28='
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jsonpickle/jsonpickle/issues/108#issuecomment-69331571>
> .
>


-- 
Marcin Tustin
Tel: +44 (0) 7773 787 105 (UK)
       +1  917 553 3974 (US)

**Comment by aldanor** [comment]
```python
>>> print(b'\x00\x00\x00\x00\x00\x00\xf0?foo'.encode('base64'))
AAAAAAAA8D9mb28=
```
The ascii part ("foo") is lost :/

**Comment by marcintustin** [comment]
>>> b'\x00\x00\x00\x00\x00\x00\xf0?foo'.encode('base64').decode('base64')
'\x00\x00\x00\x00\x00\x00\xf0?foo'

Quoted printable is probably the right thing here, though:
>>> b'\x00\x00\x00\x00\x00\x00\xf0?foo'.encode('quopri').decode('quopri')
'\x00\x00\x00\x00\x00\x00\xf0?foo'
>>> b'\x00\x00\x00\x00\x00\x00\xf0?foo'.encode('quopri')
'=00=00=00=00=00=00=F0?foo'

**Comment by aldanor** [comment]
Too many options :panda_face:

**Comment by marcintustin** [comment]
Not sure I follow?

-- 
Marcin Tustin
Tel: +44 (0) 7773 787 105 (UK)
       +1  917 553 3974 (US)

**Comment by aldanor** [comment]
base64 / quopri / latin1 / array, many options to do this -- we need another opinion :)

I personally like `quopri` encoding though since it leaves ascii as ascii and renders the rest as hex.

**Comment by marcintustin** [comment]
Just implement it as quopri. We can change it later before release.

-- 
Marcin Tustin
Tel: +44 (0) 7773 787 105 (UK)
       +1  917 553 3974 (US)

**Comment by aldanor** [comment]
Seems to work so far, I'll try and push it later tonight (this is on py3k):
```python
>>> import jsonpickle
>>> s = b'\x00\x00\x00\x00\x00\x00\xf0?foo'
>>> jsonpickle.encode(s)
'{"py/bytes": "=00=00=00=00=00=00=F0?foo"}'
>>> jsonpickle.decode(jsonpickle.encode(s))
b'\x00\x00\x00\x00\x00\x00\xf0?foo'
```

**Comment by aldanor** [comment]
The next big question is -- what do we do in Python 2? Technically, in py2 `str` *is* `bytes`, but does that mean we have to wrap every non-unicode string in `{"py/bytes": ...}`? This way, in Python 2 `jsonpickle.encode('foo')` would yield `'{"py/bytes": "foo"}'` and not just `"foo"` which may be excessively verbose.

The current behaviour in Python 2 is this (which seems quite broken):
```python
>>> jsonpickle.encode('\xff')
'"\xff"'
>>> jsonpickle.decode(jsonpickle.encode('\xff'))
ValueError: Invalid UTF-8 sequence length when decoding 'string'
```

**Comment by marcintustin** [comment]
We could forego the bytes wrapper if the encoded version is the same as the
input, but I bet there would be nasty edge cases. At the least, you'd get
different types serializing in 2 and deserializing in 3.

-- 
Marcin Tustin
Tel: +44 (0) 7773 787 105 (UK)
       +1  917 553 3974 (US)

**Comment by aldanor** [comment]
You already get different types. Even more than that, you get different types when roundtripping within Python 2:
```python
>>> jsonpickle.decode(jsonpickle.encode('foo'))
u'foo'
```

**Comment by marcintustin** [comment]
Well, then we should probably just assume that every byte string in 2 is
Unicode, unless it can't be encoded as such (exception handler?) and encode
as bytes in that case.

On Friday, January 9, 2015, Ivan Smirnov <notifications@github.com> wrote:

> You already get different types. Even more than that, you get different
> types when roundtripping within Python 2:
>
> >>> jsonpickle.decode(jsonpickle.encode('foo'))u'foo'
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jsonpickle/jsonpickle/issues/108#issuecomment-69359817>
> .
>


-- 
Marcin Tustin
Tel: +44 (0) 7773 787 105 (UK)
       +1  917 553 3974 (US)

**Comment by aldanor** [comment]
There's one catch though: http://bugs.python.org/issue6784

Python 3 actually *does* unpickle bytestrings from Python 2 as unicode strings... What were the Python core devs smoking? Which obviously makes it even more confusing, especially if we are to comply with pickle protocols. So to get/hack around this weirdness, there's a fallback option in `pickle` so you can specify something like `errors='bytes'` and hope that works (e.g. see an example [here](https://github.com/zopefoundation/zodbpickle/blob/master/README.rst#loadingstoring-python-2-strings)).

**Comment by aldanor** [comment]
One cheap way of doing this would be like so:
* In Python 2:
  * serialize `unicode` strings as is
  * for `str` strings, first try to encode them in utf-8
  * only if that fails, serialize them via the bytes wrapper
* In Python 3:
   * serialize all `str` strings as is (unicode)
   * serialize all `bytes` strings via the bytes wrapper

**Comment by marcintustin** [comment]
I'm certain that's the right thing.

**Comment by davvid** [comment]
Yeah, sounds just about right.  For str, we'd probably try to decode into unicode as utf-8, rather than decode, right?  If you use simplejson you can already see that `u'foo'` becomes `'foo'` at the lowest level.

**Comment by aldanor** [comment]
Yep, mind the typo -- decode of course, that's how it is done in #110

**Comment by aldanor** [comment]
Closed via #110
